lib/ash_json_api/serializer.ex

```diff
@@ -206,73 +220,164 @@ defmodule AshJsonApi.Serializer do
+  defp put_page_params(query, _), do: query
+
+  # Offset pagination
+
+  defp add_next_link(links, uri, query, %Ash.Page.Offset{} = paginator) do
+    %{results: results, count: count, offset: offset, limit: limit} = paginator
+
+    cond do
+      not is_nil(count) and not is_nil(offset) and offset + limit >= count ->
+        Map.put(links, :next, nil)
+
+      not is_nil(count) and is_nil(offset) and limit >= count ->
+        Map.put(links, :next, nil)
+
+      Enum.count(results) < limit ->
+        Map.put(links, :next, nil)
+
+      true ->
+        new_query =
+          query
+          |> put_page_params(next_page(paginator))
+          |> Conn.Query.encode()
+
+        link =
+          uri
+          |> put_query(new_query)
+          |> URI.to_string()
+          |> encode_link()
+
+        Map.put(links, :next, link)
+    end
   end
 
-  defp put_page_params(query, %{limit: limit}) when is_integer(limit),
-    do: Map.put(query, "page", %{limit: limit})
+  ## Cursor pagination
 
-  defp put_page_params(query, _), do: query
+  defp add_next_link(links, uri, query, %Ash.Page.Keyset{} = paginator) do
+    case paginator do
+      # This is a query for the first page, no cursors are set
+      %{results: results, after: nil, before: nil, more?: true} ->
+        new_query =
+          query
+          |> put_page_params(%{paginator | after: List.last(results).__metadata__.keyset})
+          |> Conn.Query.encode()
+
+        link =
+          uri
+          |> put_query(new_query)
+          |> URI.to_string()
+          |> encode_link()
+
+        Map.put(links, :next, link)
 
-  defp add_next_link(links, _uri, _query, %{offset: offset, limit: limit, total: total})
-       when not is_nil(total) and not is_nil(offset) and offset + limit >= total,
-       do: links
+      # Pagination forward with after, but no more results
+      %{results: [], after: _, before: nil, more?: false} ->
+        Map.put(links, :next, nil)
 
-  defp add_next_link(links, _uri, _query, %{offset: offset, limit: limit, total: total})
-       when not is_nil(total) and is_nil(offset) and limit >= total,
-       do: links
+      # Pagination forward with after, there are results
+      # If there is more, add next link, else next link is nil
+      %{results: results, after: _, before: nil, more?: true} ->
+        new_query =
+          query
+          |> put_page_params(%{paginator | after: List.last(results).__metadata__.keyset})
+          |> Conn.Query.encode()
 
-  defp add_next_link(links, uri, query, %{offset: _} = paginator) do
-    if Enum.count(paginator.results) < paginator.limit do
-      links
+        link =
+          uri
+          |> put_query(new_query)
+          |> URI.to_string()
+          |> encode_link()
+
+        Map.put(links, :next, link)
+
+      # Pagination backward, there are results
+      # Since we are paginating backwards from a result, we assume there will be more
+      %{results: results, after: nil} ->
```

Comment:
Thinking about it, is this how `next` should work? In this implementation, `next` is actually kind of behaving like `prev`, isn't it? I.e if you got a page with `before: keyset` and then did `next` you'd get back the page you were just on. I think `next` page refer to the remaining stuff before the now-first thing on the page.

Reply:
I might be understanding this wrong or this might be the right way, but I was looking at the examples from here:
[JSON API paginating](https://jsonapi.org/profiles/ethanresnick/cursor-pagination/#auto-id-query-parameters) and it seems regardless of the direction, they set the next as after the before, and prev as before the before(:)). 

I guess with a simple use case example, suppose I have collection `[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]`.
Looking at it from a ui perspective I would have prev disabled/nil and next cursor set to `5`.
Once I have paginated to `5`, I would have two cursor links, `prev at 6` and `next at 10`.
I would then take the `prev at 6` cursor and say give me all elements before `6` which will be `[1, 2, 3, 4, 5]` (Resorted).  
The current cursors would then be `prev at 1` and `next at 5`.

I think the use case you are describing is better achieved by applying a sort? What do you think?

Reply:
I think I see what you're saying. Lets go with it :) Keyset pagination is broken as is, so if anyone has issues we can adjust the behavior :D